src/lib/config.ts

```diff
@@ -0,0 +1,417 @@
+
+    return {
+      secret,
+      url: this.url,
+      graphqlHost: this.graphqlHost,
+      graphqlPort: this.graphqlPort,
+    };
+  }
+}
+
+class Stack {
+  /**
+   * The endpoint name to use as a base.
+   */
+  endpoint: string;
+  /**
+   * The database path to use.
+   */
+  database: string | undefined;
+
+  constructor(config: Config) {
+    this.endpoint = config.str("endpoint");
+    this.database = config.strOpt("database");
+  }
+
+  updateDatabase(database: string | undefined) {
+    if (database !== undefined) {
+      this.database = database;
+    }
+  }
+
+  makeSecret(endpoint?: Endpoint, scope?: string, role?: string) {
+    const scope0 = scope ?? this.database;
+    const secret =
+      endpoint?.secret + (scope0 ? `:${scope0}` : "") + ":" + (role ?? "admin");
+
+    return secret;
+  }
+}
+
+const getURLFromConfig = (config: Config): string | undefined => {
+  const url = config.strOpt("url");
+  const scheme = config.strOpt("scheme");
+  const domain = config.strOpt("domain");
+  const port = config.numberOpt("port");
+
+  if (
+    url === undefined &&
+    (domain !== undefined || port !== undefined || scheme !== undefined)
+  ) {
+    const scheme0 = scheme ?? "https";
+    const domain0 = domain ?? "db.fauna.com";
+    const port0 = port ? `:${port}` : "";
+    return `${scheme0}://${domain0}${port0}`;
+  } else {
+    return url;
+  }
+};
+
+/**
+ * Builds the options provided to the faunajs client.
+ * Tries to load the ~/.fauna-shell file and read the default endpoint from there.
+ *
+ * Assumes that if the file exists, it would have been created by fauna-shell,
+ * therefore it would have a defined endpoint.
+ *
+ * Flags like --host, --port, etc., provided by the CLI take precedence over what's
+ * stored in ~/.fauna-shell.
+ *
+ * The --endpoint flag overries the default endpoint from fauna-shell.
+ *
+ * If ~/.fauna-shell doesn't exist, tries to build the connection options from the
+ * flags passed to the script.
+ *
+ * It always expect a secret key to be set in ~/.fauna-shell or provided via CLI
+ * arguments.
+ */
+export const lookupEndpoint = (flags: any, scope: string, role: string) => {
+  const rootConfig = ini.parse(readFileOpt(getRootConfigFile()));
+  const localConfigPath = getLocalConfigFile();
+  const localConfig = localConfigPath
+    ? ini.parse(fs.readFileSync(localConfigPath, "utf8"))
+    : undefined;
+
+  return lookupEndpoint0({
+    flags,
+    rootConfig,
+    localConfig,
+    scope,
+    role,
+  });
+};
+
+export const lookupEndpoint0 = (opts: {
+  flags?: { [key: string]: any };
+  rootConfig?: { [key: string]: any };
+  localConfig?: { [key: string]: any };
+  scope?: string;
+  role?: string;
+}): {
+  secret: string;
+  url: string;
+  graphqlHost: string;
+  graphqlPort: number;
+} => {
+  const flags = new Config("flag", opts.flags ?? {});
+
+  const rootConfig = new RootConfig(
+    new Config("config key", opts.rootConfig ?? {})
```

Comment:
nit: this obviously makes no difference at all... but I'd feel a bit better if our keys for these were something like `rootConfig` and `projectConfig`

Reply:
the keys there show up in errors. So for "config key", you'd see this error:
```
Expected string for config key url, got number
```
and for "flag" you'd see
```
Expected string for flag url, got number
```

so we could definitely update them to "root config key" and "project config key" or whatnot, but I don't think "rootConfig" and "projectConfig" would make that much sense.